src/tests/fuzzing/fuzz_pkcs15_decode.c

```diff
@@ -20,56 +20,80 @@
+	static sc_card_t *card = NULL;
+	struct sc_pkcs15_tokeninfo *tokeninfo = NULL;
+	int (* decode_entries[])(struct sc_pkcs15_card *, struct sc_pkcs15_object *,
+			const u8 **nbuf, size_t *nbufsize) = {
+		sc_pkcs15_decode_prkdf_entry, sc_pkcs15_decode_pukdf_entry,
+		sc_pkcs15_decode_skdf_entry, sc_pkcs15_decode_cdf_entry,
+		sc_pkcs15_decode_dodf_entry, sc_pkcs15_decode_aodf_entry
+	};
+	int algorithms[] = { SC_ALGORITHM_RSA, SC_ALGORITHM_EC, SC_ALGORITHM_GOSTR3410, SC_ALGORITHM_EDDSA };
 
-    if (!ctx)
-        sc_establish_context(&ctx, "fuzz");
-    if (!p15card) {
-        card.ctx = ctx;
-        p15card = sc_pkcs15_card_new();
-        if (p15card) {
-            p15card->card = &card;
-        }
-    }
+	/* Establish context for fuzz app*/
+	sc_establish_context(&ctx, "fuzz");
+	if (!ctx)
+		return 0;
 
-    for (i = 0; i < sizeof decode_entries/sizeof *decode_entries; i++) {
-        struct sc_pkcs15_object *obj;
-        const u8 *p = Data;
-        size_t len = Size;
-        obj = calloc(1, sizeof *obj);
-        while (SC_SUCCESS == decode_entries[i](p15card, obj, &p, &len)) {
-            sc_pkcs15_free_object(obj);
-            obj = calloc(1, sizeof *obj);
-        }
-        sc_pkcs15_free_object(obj);
-    }
+	if (fuzz_connect_card(ctx, &card, &reader, Data, Size) != SC_SUCCESS)
+		goto err;
 
-    struct sc_pkcs15_pubkey *pubkey = calloc(1, sizeof *pubkey);
-    sc_pkcs15_decode_pubkey(ctx, pubkey, Data, Size);
-    sc_pkcs15_free_pubkey(pubkey);
+	sc_pkcs15_bind(card, NULL, &p15card);
+	if (!p15card)
+		goto err;
 
-    struct sc_pkcs15_tokeninfo *tokeninfo = sc_pkcs15_tokeninfo_new();
-    sc_pkcs15_parse_tokeninfo(ctx, tokeninfo, Data, Size);
-    sc_pkcs15_free_tokeninfo(tokeninfo);
+	for (i = 0; i < sizeof decode_entries/sizeof *decode_entries; i++) {
+		struct sc_pkcs15_object *obj;
+		const u8 *p = NULL;
+		size_t len = 0;
+		if (!(obj = calloc(1, sizeof *obj)))
+			goto err;
+		fuzz_get_chunk(reader, &buf, &buf_len);
+		p = buf;
+		len = (size_t) buf_len;
+		while (SC_SUCCESS == decode_entries[i](p15card, obj, &p, (size_t *) &len)) {
+			sc_pkcs15_free_object(obj);
+			if (!(obj = calloc(1, sizeof *obj)))
+				goto err;
+		}
+		sc_pkcs15_free_object(obj);
+	}
 
-    sc_pkcs15_parse_unusedspace(Data, Size, p15card);
+	fuzz_get_chunk(reader, &buf, &buf_len);
+	for (i = 0; i < 4; i++) {
+		struct sc_pkcs15_pubkey *pubkey = calloc(1, sizeof *pubkey);
+		if (!pubkey)
+			goto err;
+		pubkey->algorithm = algorithms[i];
+		sc_pkcs15_decode_pubkey(ctx, pubkey, buf, buf_len);
+		sc_pkcs15_free_pubkey(pubkey);
+	}
 
-    return 0;
+	fuzz_get_chunk(reader, &buf, &buf_len);
+	tokeninfo = sc_pkcs15_tokeninfo_new();
+	sc_pkcs15_parse_tokeninfo(ctx, tokeninfo, buf, buf_len);
+	sc_pkcs15_free_tokeninfo(tokeninfo);
+
+	fuzz_get_chunk(reader, &buf, &buf_len);
+	sc_pkcs15_parse_unusedspace(buf, buf_len, p15card);
+
+	sc_pkcs15_card_free(p15card);
+
+err:
+	sc_disconnect_card(card);
+	sc_release_context(ctx);
+	return 0;
 }
```

Comment:
For the encoding/decoding operations that are independent from each other, it would maybe be better to either use separate fuzzing tests for them or to *not* advance the buffer with the fuzzing input (`buf`/`buf_len` before `sc_pkcs15_parse_tokeninfo`, `sc_pkcs15_parse_unusedspace`, for example). This allows a wider range of fuzzing input to be useful for actual testing...